### PR TITLE
Convert shared CFG construction library to a parameterized module

### PR DESCRIPTION
**Pull request by hvitved**
This PR converts the existing `ControlFlowGraphImplShared.qll` to a parameterized module, and puts it into a new `controlflow` QL pack (where, eventually, we can also put stuff like basic blocks and guards). In this PR, only Ruby is adopting the new shared library; I plan to do C# follow-up, and perhaps Swift (unless @github/codeql-swift beats me to it). 

Whitespace-insensitive commit-by-commit review is suggested.

~The PR makes use of `extends` inside parameterized modules, which is only publicly supported in 2.14.0 (expected to be released July 11th), so cannot be merged until then (this is the reason why the CI check `Compile all queries using the latest stable CodeQL CLI` currently fails).~ 2.14.0 has now been released, so this PR is safe to merge.

### CODE REVIEW

File: ruby/ql/lib/codeql/ruby/controlflow/CfgNodes.qll [hunk 1]
```diff
@@ -507,9 +449,11 @@
 
   /** A control-flow node that wraps a `CasePattern`. */
   class CasePatternCfgNode extends AstCfgNode {
-    override string getAPrimaryQlClass() { result = "CasePatternCfgNode" }
+    CasePattern e;
```
[the remark this answers is not in the document]
Comment (hvitved) [reply ruby/ql/lib/codeql/ruby/controlflow/CfgNodes.qll:454]: FP, fixed by https://github.com/github/codeql/pull/13517.

File: ruby/ql/lib/codeql/ruby/controlflow/ControlFlowGraph.qll [hunk 2]
```diff
@@ -33,60 +33,39 @@
...
-
-  /** Gets the scope of this node. */
-  final CfgScope getScope() { result = getNodeCfgScope(this) }
-
-  /** Gets the basic block that this control flow node belongs to. */
-  BasicBlock getBasicBlock() { result.getANode() = this }
-
   /** Gets a successor node of a given type, if any. */
-  final CfgNode getASuccessor(SuccessorType t) { result = getASuccessor(this, t) }
+  final CfgNode getASuccessor(SuccessorType t) { result = super.getASuccessor(t) }
```
[the remark this answers is not in the document]
Comment (hvitved) [reply ruby/ql/lib/codeql/ruby/controlflow/ControlFlowGraph.qll:47]: FP, should be fixed by https://github.com/github/codeql/pull/13521.

File: shared/controlflow/codeql/controlflow/Cfg.qll [hunk 3]
```diff
@@ -0,0 +1,1282 @@
...
+
+  /**
+   * A control flow node.
+   *
+   * A control flow node is a node in the control flow graph (CFG). There is a
+   * many-to-one relationship between CFG nodes and AST nodes.
+   *
+   * Only nodes that can be reached from an entry point are included in the CFG.
+   */
+  abstract private class NodeImpl extends TNode {
```
[the remark this answers is not in the document]
Comment (hvitved) [reply shared/controlflow/codeql/controlflow/Cfg.qll:981]: FP, should be fixed by https://github.com/github/codeql/pull/13521.

File: shared/controlflow/codeql/controlflow/Cfg.qll [hunk 4]
```diff
@@ -0,0 +1,1293 @@
...
+  class ControlFlowTreeBase {
+    /** Gets a textual representation of this element. */
+    string toString();
+
+    /** Gets the location of this element. */
+    Location getLocation();
+  }
+
+  /** An AST node. */
+  class AstNode extends ControlFlowTreeBase;
```
Comment (aibaars) [inline shared/controlflow/codeql/controlflow/Cfg.qll:20]: Do we really need `AstNode` and `ControlFlowTreeBase` in the signature?

File: shared/controlflow/codeql/controlflow/Cfg.qll [hunk 5]
```diff
@@ -0,0 +1,1293 @@
...
+module Make<LocationSig Location, InputSig<Location> Input> {
+  private import Input
+
+  final private class ControlFlowTreeBaseFinal = ControlFlowTreeBase;
+
+  /** An element with associated control flow. */
+  abstract class ControlFlowTree extends ControlFlowTreeBaseFinal {
+    /** Holds if `first` is the first element executed within this element. */
+    pragma[nomagic]
+    abstract predicate first(AstNode first);
```
Comment (aibaars) [inline shared/controlflow/codeql/controlflow/Cfg.qll:136]: Would it be possible to replace most uses of `AstNode` with `ControlFlowTree(Base)` , something like:
```suggestion
    abstract predicate first(ControlFlowTreeBase first);
```

File: shared/util/codeql/util/Location.qll [hunk 6]
```diff
@@ -0,0 +1,35 @@
+/** Provides classes for working with locations. */
+
+/**
+ * A location as given by a file, a start line, a start column,
+ * an end line, and an end column.
+ *
+ * For more information about locations see [Locations](https://codeql.github.com/docs/writing-codeql-queries/providing-locations-in-codeql-queries/).
+ */
+signature class LocationSig {
+  /** Gets the 1-based line number (inclusive) where this location starts. */
```
Comment (aibaars) [inline shared/util/codeql/util/Location.qll:10]: I guess you "forgot" `getFile` on purpose, right?

**Review by aibaars — COMMENTED** [review]
Generally looks good to me. I wonder if we can get rid of  most uses of `AstNode`. Also don't forget to include the new library in the list of things to release in the `build` file.

File: shared/controlflow/codeql/controlflow/Cfg.qll (hunk 4 above)
Comment (hvitved) [reply shared/controlflow/codeql/controlflow/Cfg.qll]: No, you are right, I have merged them.

File: shared/util/codeql/util/Location.qll (hunk 6 above)
Comment (hvitved) [reply shared/util/codeql/util/Location.qll:10]: Yes, because that would require a `File` class in scope.

**Review by MathiasVP — APPROVED** [review]
LGTM once DCA comes back happy 🤞!

**Pull request merged**